proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+ghci> :kind! LiftedCtorToSymbol 'Data.Either.Left
+LiftedCtorToSymbol 'Data.Either.Left :: Symbol
+= "Left"
+```
+
+## Proposed Library Change Specification
+
+- `compactAddHollow#`, `reifyInfoTablePtr#`, and `InfoTablePtrOf#` are exposed in `GHC.Prim`
+- `LiftedCtorToSymbol` is exposed in `GHC.TypeLits`
+
+## Examples
+
+### Primitive usage
+
+#### Allocate a hollow `Left` constructor in region `compactRegion` and complete it with a `()` constructor
+
+```haskell
+example :: Compact# -> State# RealWorld -> (# State# RealWorld, Either () () #)
+example compactRegion s0 =
+  let leftInfoPtr :: Addr#
+      leftInfoPtr = unsafeCoerceAddr (reifyInfoTablePtr# (# #) :: InfoTablePtrOf# 'Left)
+      unitInfoPtr :: Addr#
+      unitInfoPtr = unsafeCoerceAddr (reifyInfoTablePtr# (# #) :: InfoTablePtrOf# '())
+   in case compactAddHollow# compactRegion leftInfoPtr s0 of
+        (# s1, leftCtor, headerSize, wordSize #) -> case anyToAddr# leftCtor s1 of
+          (# s2, pLeftCtor #) -> case compactAddHollow# compactRegion unitInfoPtr s2 of
+            (# s3, unitCtor, headerSize, wordSize #) -> case anyToAddr# unitCtor s3 of
+              (# s4, pUnitCtor #) -> case writeAddrOffAddr# pLeftCtor (word2Int# headerSize) pUnitCtor s4 of
+                s5 -> (# s5, leftCtor #)
+```
+
+First we start by reifying the info table pointer for the two constructors we'll need, `Left` and `()`.
+
+Then we allocate a hollow `Left` constructor in the compact region, and get a pointer to it. We do the same for a `()` constructor.
+
+Finally we set the (only) field of the `Left` constructor, located at offset `headerSize` from base address of that constructor object, to point to the `()` constructor, using existing primop `writeAddrOffAddr#`.
+
+The returned object of type `Either () Int` can be read safely as it no longer has any unspecified field.
+
+#### Allocate hollow `stg_IND` object in region `compactRegion`
+
+```haskell
+let indInfoPtr = unsafeCoerceAddr (reifyInfoTablePtr# (# #) :: InfoTablePtrOf# "IND")
+ in compactAddHollow# compactRegion indInfoPtr state
+```
+
+As in the previous example, we reify the info table pointer for the `stg_IND` object, and then allocate a hollow `stg_IND` object in the compact region. Same can be done with other special STG objects.
+
+#### Get info about fields of a data constructor using `GHC.Generics` and `LiftedCtorToSymbol`
+
+Here I present how to extract information about a data constructor (number and type of its fields, among others) in a type-level structure, starting from its type-level (lifted) representation and the type `a` that it belongs to. The type should implement `Generic` (i.e. `Rep a ()` should reduce).
+
+This type-level machinery first use `LiftedCtorToSymbol` to get the name of the specified data constructor as a `Symbol`, and then uses it to filter the generic representation `Rep a ()`.
+
+We assume existence of type-level `(++)`, `(<|>)` and `FromJust :: Maybe k -> k operators`
```

Comment:
I think I know what you intend for these to mean, but you can't assume that the reader knows.